packages/grafana-data/src/types/datasource.ts

```diff
@@ -362,6 +362,12 @@ abstract class DataSourceApi<
    * @alpha -- experimental
    */
   getDefaultQuery?(app: CoreApp): Partial<TQuery>;
+
+  /**
+   * Define if the contents of a query are complete
+   * @alpha -- experimental
+   */
+  isCompleteQuery?(query?: DataQuery): boolean;
```

Comment:
What about `isValidQuery`?

Reply:
It's kind of a tough call to make but we want to make a distinction between completeness and validity here. We dont want to save empty queries because theres nothing to save. That is a coupling we want to have between the query and and the query history logic.

In contrast, we want to save complete but invalid queries because someone might be trying to debug something, also in some cases a query that is saved might become invalid. We don't want query history to know or care about a query's validity.

Also, I feel like completeness is simpler logic that should be able to be checked with a little client side logic. Validity is something more tied to the datasource and would require a server call. 

Does that make sense?

Reply:
Would it make sense to make that method very specific to the query history? I could think about some use cases where - as a data source developer - some queries should be hidden from the history even though they are not empty. I was thinking about the way it's done for the inspector, that is just using the `hideFromInspector` flag for that:
https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/datasource.ts#L570

Reply:
@svennergr I think we could build off of what is already there for `filterQuery` for this - it has other uses other than query history